### ISSUE TITLE
update CREATE CLUSTER syntax to always create a default replica

### ISSUE THREAD

**Issue by JLDLaughlin** [MaterializeInc/materialize#12736]
### Feature request

We recently landed some new syntax in Materialize: `CREATE CLUSTER` and `CREATE CLUSTER REPLICA`. Given internal feedback that we've received in our Cloud testing sessions, I want to suggest changing this syntax.

### What we have now

Clusters are just logical groupings of dataflow-powered objects. Users will need to create a cluster and then some number of replicas in order to actually use Materialize.

In the Cloud product, we create a default cluster with a default replica for the user. I think this is great! But this also means the second cluster users create won't have a replica by default, making the cluster/replica experience asymmetric.

### What's the problem

I think the cluster/replica concept is a lot of mental overhead for new users, particularly since they will be the first concepts they encounter in Materialize. In my opinion, we want to skip people ahead to materialized views as quickly as we can. Making them understand clusters/replicas feels like the wrong way to spend users' limited time and attention. 

### My proposal

@ruf-io suggested this change, which I like!

```
CREATE CLUSTER dev; -- Creates a cluster with 1 replica size xsmall

CREATE CLUSTER stage (SIZE = 'small'); -- Creates a cluster with 1 replica size small
```

Basically, clusters should always create a default replica. The default replica will always be xsmall, but users can opt for a larger size.

**Comment by JLDLaughlin** [comment]
cc @ruf-io @morsapaes @joacoc @heeringa @benesch @sploiselle

**Comment by heeringa** [comment]
I endorse this proposal.   Here is why:

* Most users will create a cluster with the immediate intention of creating physical resources -- at least one replica -- to do the work so this combines two steps into one in a natural way.
* The convention of cluster being a shell is not intuitive -- my sense is that the proposed syntax better matches the default mental model of our customers.  I might be wrong.
* The convention of cluster being a shell makes for some challenging documentation -- to accurately describe it requires folks to pivot to more conceptual thinking.  This proposal will make our documentation simpler and cleaner.

**Comment by benesch** [comment]
I've got to run, but I want to flag that I am pretty firmly opposed to this! I can expand upon this in more detail lately, but I feel strongly that the syntax needs to correspond to the core concepts as directly as possible. If the core concepts are not intuitive, then we either need to simply those concepts, or invest in the user education required to get users on board with our model.

Trying to hide complexity via SQL syntax tends to create bigger problems later. The classic example of this is set returning functions. PostgreSQL allows you to put SRFs in the `SELECT` list even though it semantically makes no sense, e.g. `SELECT generate_series(1, 2)`. This works well for simple cases, but is completely nonsensical in complicated cases (e.g., what should `SELECT generate_series(generate_series(1, 2), 2), generate_series(1, 2)` return?). 

My fear is that attempting to simplify the UX of creating a cluster by papering over the complexity of replicas will create extremely painful problems for power users down the line.

> * The convention of cluster being a shell is not intuitive -- my sense is that the proposed syntax better matches the default mental model of our customers. I might be wrong.
> * The convention of cluster being a shell makes for some challenging documentation -- to accurately describe it requires folks to pivot to more conceptual thinking. This proposal will make our documentation simpler and cleaner.

This actually strikes me as an argument to not adjust the syntax. The existing syntax matches the reality of how Materialize works—as it should! 

--- 

I think the right place to invest in simplicity here is in the web UI. It's _way_ easier to build visual UIs that serve both beginners and experts than it is with SQL. I have a mockup in Slack that I'll try to track down.

I'd also note that we have not yet had any documentation of the SQL commands! It's hard to give the existing syntax a fair shake until the docs are published.

**Comment by JLDLaughlin** [comment]
So I think we might be almost agreeing? I like that UI! But it doesn't let you create a cluster without a replica. I just want the SQL syntax to do the same thing!

**Comment by benesch** [comment]
> But it doesn't let you create a cluster without a replica.

It would! If you set the number of replicas to zero. 

Part of the tension here is that the SQL interface needs to serve both machines and humans. Tools that we build on top of the SQL layer will autogenerate the `CREATE CLUSTER` and `CREATE CLUSTER REPLICA` statements, like an eventual Terraform provider. Those tools very much want clusters to start empty, rather than starting with a default replica that they then have to move.

Also, I want to flag that the long term plan is to move to an autoscaling, declarative API here. Imagine:

```
CREATE CLUSTER x, REPLICATION FACTOR 3, MIN SIZE ‘small’, MAX SIZE ‘large’, IDLE SHUTDOWN ‘3 minutes’
``` 

That would create a cluster that targets three healthy replicas, auto scales them from small to large as load demands, and shuts them down if no queries have arrived in 3 minutes. 

The current syntax is a bit of a stopgap until we can build the smarts into Materialize to do that automation. I think that long term target likely satisfies everyone?

Is that helpful? The meta point, I think, is that I would much rather invest in switching our model of replica management from imperative to declarative than try to paper over the deficiencies of our current model with syntax sugar. (I call our current API “imperative” because it requires you to type `CREATE CLUSTER REPLICA` and explicitly name and size your replicas. The proposed long-term API is “declarative” because you declare your desired replication factor and the system handles the rest; you never have to explicitly name or size individual replicas.) Introducing that simple syntax now will impair our ability to use that simple syntax for the good declarative API, once we’ve got the smarts to do the automatic scaling/replication in Materialize!

**Comment by heeringa** [comment]
Playing off your desired syntax above, what if the `CREATE CLUSTER` command now supported creating both a single replica of a particular size and creating a logical container.

`CREATE CLUSTER stage (SIZE = 'large'); -- Creates a cluster with 1 replica size large`
`CREATE CLUSTER stage; -- Creates a cluster with with 1 replica size extra small`
`CREATE EMPTY CLUSTER stage; -- Creates an empty cluster representing a logical container`

Or something where our idea of a CLUSTER is a logical container but the default syntax for a cluster matches people's mental models of having at least one replica?

**Comment by benesch** [comment]
> Playing off your desired syntax above, what if the `CREATE CLUSTER` command now supported creating both a single replica of a particular size and creating a logical container.
> 
> `CREATE CLUSTER stage (SIZE = 'large'); -- Creates a cluster with 1 replica size large` `CREATE CLUSTER stage; -- Creates a cluster with with 1 replica size extra small` `CREATE EMPTY CLUSTER stage; -- Creates an empty cluster representing a logical container`
> 
> Or something where our idea of a CLUSTER is a logical container but the default syntax for a cluster matches people's mental models of having at least one replica?

I'm afraid this is exactly what I think we need to avoid! Attempting to paper over inherent complexity with SQL syntax sugar has gotten both us and PostgreSQL into trouble in the past. I talked a bit about PostgreSQL's mistake with allowing set-returning functions in scalar position above. The (arguable) mistake we made in Materialize was allowing 

```
CREATE MATERIALIZED VIEW v AS SELECT ...;
```

to be syntax sugar for:

```
CREATE VIEW v AS SELECT ...;
CREATE DEFAULT INDEX ON v;
```

The syntax leads users to believe that there are two types of views, materialized and unmaterialized, and you have to choose between them when you create the view. And indeed this is how things work in other databases. But it is emphatically not how things work in Materialize—you can create and drop indexes to materialize or unmaterialize a view at any point! The situation is even worse now with clusters, where `CREATE MATERIALIZED VIEW` creates a view that is materialized _only_ in the active cluster.

It also causes other commands to fail to work as users expect. `SHOW CREATE VIEW` "loses" the `MATERIALIZED` keyword (MaterializeInc/materialize#3404, MaterializeInc/materialize#5703). You can't type `ALTER VIEW v SET MATERIALIZED = false` to unmaterialize a view; you have to instead learn to type `DROP INDEX v_primary_idx`.

Having different syntax for creating an empty vs non-empty cluster is going to lead to the same sort of confusion. Are `EMPTY` clusters different from non-empty clusters? Why can't you type `ALTER CLUSTER ... SET EMPTY = true/false` ?

----

To back up for a second, I think there are a few separate concerns here that we should tease apart:

* **Teaching users who are new to Materialize about clusters/replicas is a lot of overhead.**

  I think we're a bit overfocused on clusters and replicas right now because they are the only new user-facing feature of Materialize Platform. But I think the answer here is to temper our excitement about clusters in the new user flow, and do what Jessica has proposed in MaterializeInc/cloud#3152!

* **`CREATE CLUSTER foo` is a footgun that creates an empty cluster.**

  I think we should move in the direction of requiring _more_ explicitness rather than less—I'd rather we require users to specify the replicas and their sizes explicitly. I'll file a follow-up issue about this shortly. 

* **Manually manage cluster replicas is annoying and an educational burden.**

  This we fix by introducing automatic replica management via the declarative API I mentioned above. Then you'll be able to set `MIN SIZE`/`MAX SIZE`/`REPLICATION FACTOR` on a cluster directly, and never have to think about individual replicas. But this isn't as easy as a simple syntax change—we've got to build the auto-management smarts into Materialize. And in fact adding syntax sugar now may prevent us from using that simpler syntax for automatic replica management, once we have it.

**Comment by benesch** [comment]
To phrase one of my concerns perhaps a bit more succinctly: I don't think we should support

```
CREATE CLUSTER ... SIZE 'small'
```

until we can also support:

```
ALTER CLUSTER ... SIZE 'medium' -- Size up cluster.
ALTER CLUSTER ... REPLICATION FACTOR 3 -- Increase availability of cluster.
```

The goal is very much to get there, but there is a lot of work required to support those `ALTER` statements!

**Comment by heeringa** [comment]
Your proposal in https://github.com/MaterializeInc/materialize/issues/12769:

```
CREATE CLUSTER foo REPLICAS ();
CREATE CLUSTER foo REPLICAS (
    r1 (SIZE 'small')
);
CREATE CLUSTER foo REPLICAS (
    r1 (SIZE 'small'),
    r2 (SIZE 'small'),
)
```

makes me much happier because it makes it explicitly clear that the cluster will have zero replicas; we'll still need to educate our users that by 'replica' we really mean 'compute instance' because the first replica is not replicating *anything*.

**Comment by benesch** [comment]
> Your proposal in #12769 [...] makes me much happier because it makes it explicitly clear that the cluster will have zero replicas

Great!

> we'll still need to educate our users that by 'replica' we really mean 'compute instance'

I'm not enamored of the term "compute instance", but given that it exists in our codebase/formalism I do want to note that a compute instance === a cluster, according to Frank's definition. I think we're probably best avoiding the term "instance" in any external comms because it's real confusing.

----

I have a hunch that a lot of the confusion is stemming from the difference between passive/active replication. In passive replication, there is a primary replica that receives all client requests and forwards requests to the secondary/backup replicas. In active replication, there are any number of replicas, all of which are identical to one another, and clients broadcast requests to all replicas. 

I think a lot of folks are expecting passive replication terminology, but Materialize is an active replication system! IME, with active replication systems it is normal and familiar to talk about the degenerate case of having only one replica.

**Comment by heeringa** [comment]
I know that compute instance is equivalent to cluster in Frank's world, but I have a hard time accepting the word "instance" being use to denote an empty cluster because, well, nothing is actually _instantiated_.  

The world I'm willing to live in is that we have clusters, which may be empty.  These are logical compute structures.
We have active replicas, which represent physical computing resources.

I'm set with that terminology and can stand behind it and the proposed syntax!

**Comment by heeringa** [comment]
Resolution to this proposal is: https://github.com/MaterializeInc/materialize/issues/12769

```
CREATE CLUSTER foo REPLICAS ();
CREATE CLUSTER foo REPLICAS (
    r1 (SIZE 'small')
);
CREATE CLUSTER foo REPLICAS (
    r1 (SIZE 'small'),
    r2 (SIZE 'small'),
)
```
